plugins/thunderbolt/fu-plugin-thunderbolt.c

```diff
@@ -0,0 +1,736 @@
+		g_warning ("failed to get Device ID: %s", error->message);
+		return;
+	}
+
+	is_host = fu_plugin_thunderbolt_is_host (device);
+	vendor_id = g_strdup_printf ("TBT:0x%04X", (guint) vid);
+	device_id = g_strdup_printf ("TBT-%04x%04x", (guint) vid, (guint) did);
+
+	dev = fu_device_new ();
+	fu_device_set_id (dev, uuid);
+
+	fu_device_set_metadata (dev, "sysfs-path", devpath);
+	name = g_udev_device_get_sysfs_attr (device, "device_name");
+	if (name != NULL) {
+		if (is_host) {
+			g_autofree gchar *pretty_name = NULL;
+			pretty_name = g_strdup_printf ("%s Thunderbolt Controller", name);
+			fu_device_set_name (dev, pretty_name);
+		} else {
+			fu_device_set_name (dev, name);
+		}
+	}
+
+	vendor = g_udev_device_get_sysfs_attr (device, "vendor_name");
+	if (vendor != NULL)
+		fu_device_set_vendor (dev, vendor);
+	fu_device_set_vendor_id (dev, vendor_id);
+	fu_device_add_guid (dev, device_id);
+	version = g_udev_device_get_sysfs_attr (device, "nvm_version");
+	if (version != NULL)
+		fu_device_set_version (dev, version);
+	fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_ALLOW_ONLINE);
+	fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_ALLOW_OFFLINE);
+	if (is_host)
+		fu_device_add_flag (dev, FWUPD_DEVICE_FLAG_INTERNAL);
+
+	fu_plugin_cache_add (plugin, id, dev);
+	fu_plugin_device_add_delay (plugin, dev);
+}
+
+static void
+fu_plugin_thunderbolt_remove (FuPlugin *plugin, GUdevDevice *device)
+{
+	FuDevice *dev;
+	g_autofree gchar *id = NULL;
+
+	id = fu_plugin_thunderbolt_gen_id (device);
+	dev = fu_plugin_cache_lookup (plugin, id);
+	if (dev == NULL)
+		return;
+
+	fu_plugin_cache_remove (plugin, id);
+	fu_plugin_device_remove (plugin, dev);
+}
+
+static void
+fu_plugin_thunderbolt_change (FuPlugin *plugin, GUdevDevice *device)
+{
+	FuDevice *dev;
+	const gchar *version;
+	g_autofree gchar *id = NULL;
+
+	id = fu_plugin_thunderbolt_gen_id (device);
+	dev = fu_plugin_cache_lookup (plugin, id);
+	if (dev == NULL) {
+		g_warning ("got change event for unknown device, adding instead");
+		fu_plugin_thunderbolt_add (plugin, device);
+		return;
+	}
+
+	fu_plugin_device_remove (plugin, dev);
+	version = g_udev_device_get_sysfs_attr (device, "nvm_version");
+	fu_device_set_version (dev, version);
+}
+
+static gboolean
+udev_uevent_cb (GUdevClient *udev,
+		const gchar *action,
+		GUdevDevice *device,
+		gpointer     user_data)
+{
+	FuPlugin *plugin = (FuPlugin *) user_data;
+	FuPluginData *data = fu_plugin_get_data (plugin);
+
+	if (action == NULL)
+		return TRUE;
+
+	g_debug ("uevent for %s: %s", g_udev_device_get_sysfs_path (device), action);
+
+	if (data->update_notify != NULL) {
+		g_debug ("using update notify handler for uevent");
+
+		data->update_notify (plugin, device, action, data->update_data);
+		return TRUE;
+	}
+
+	if (g_str_equal (action, "add")) {
+		fu_plugin_thunderbolt_add (plugin, device);
+	} else if (g_str_equal (action, "remove")) {
+		fu_plugin_thunderbolt_remove (plugin, device);
+	} else if (g_str_equal (action, "change")) {
```

Comment:
I was wondering this and was also pretty convinced (from a glance of the kernel patches) that it will never happen. It doesn't hurt to have it though. No big deal to remove either though. ;)